### ISSUE TITLE
[gatsby-plugin-mdx] Plugin has generated no Gatsby nodes

### ISSUE THREAD

**Issue by yamalight** [gatsbyjs/gatsby#15594]
## Description

`gatsby-plugin-mdx` generates no nodes for GraphQL.
I assume this is a bug since it looks like it _should_ generate them, but it's not documented anywhere - if that's not the case, please close the issue.

### Steps to reproduce

Here's a repo with basic project: https://github.com/yamalight/gatsby-mdx-bug-demo

Steps to re-create:

1. Create new gatsby project from default starter
2. Install `gatsby-plugin-mdx` and add it to config
3. Move `src/pages/index.js` to `src/pages/index.mdx` and remove component export
4. Run `npm start`

### Expected result

Querying GraphQL endpoint with the following query should return 1 result:

```
query MyQuery {
  allMdx {
    totalCount
    edges {
      node {
        id
      }
    }
  }
}
```

### Actual result

The query returns zero result, but the page actually renders fine and is present in `allSitePage`.

Note: during `npm start` gatsby shows the following warning in the console:

```
warn The gatsby-plugin-mdx plugin has generated no Gatsby nodes. Do you need it?
```

### Environment

Note: I've also tried it on Debian machine with Node 12.6.0 with same result.

  System:
    OS: macOS 10.14.5
    CPU: (8) x64 Intel(R) Core(TM) i7-4850HQ CPU @ 2.30GHz
    Shell: 5.3 - /bin/zsh
  Binaries:
    Node: 12.5.0 - /usr/local/bin/node
    Yarn: 1.16.0 - /usr/local/bin/yarn
    npm: 6.10.0 - /usr/local/bin/npm
  Languages:
    Python: 2.7.16 - /usr/local/bin/python
  Browsers:
    Chrome: 75.0.3770.100
    Safari: 12.1.1
  npmPackages:
    gatsby: ^2.13.12 => 2.13.12 
    gatsby-image: ^2.2.4 => 2.2.4 
    gatsby-plugin-manifest: ^2.2.1 => 2.2.1 
    gatsby-plugin-mdx: ^1.0.9 => 1.0.9 
    gatsby-plugin-offline: ^2.2.1 => 2.2.1 
    gatsby-plugin-react-helmet: ^3.1.0 => 3.1.0 
    gatsby-plugin-sharp: ^2.2.3 => 2.2.3 
    gatsby-source-filesystem: ^2.1.3 => 2.1.3 
    gatsby-transformer-sharp: ^2.2.1 => 2.2.1

**Comment by yamalight** [comment]
Okay, I think I've figured it out - if I add filesystem source that processes pages before mdx plugin - it starts working.
Is that expected behavior? 🤔

**Comment by jonniebigodes** [comment]
@yamalight in a nutshell yes. Like so many other plugins `gatsby-plugin-mdx` needs a auxiliary plugin so that it can look for the files and then handle them internally, which in turn can be parsed or for instance transformed and shown. For me personnally when dealing with files i like to put the `gatsby-source-filesystem` at the top of the plugin list so that they are handled first.  

Feel free to provide feedback so that we can close this issue or continue to work on it till we find a suitable solution.

**Comment by yamalight** [comment]
@jonniebigodes Since the plugin just works for `.mdx` files - I assumed it'd also create GraphQL nodes for them. I feel like it'd be great to document that this is not the case and filesystem plugin is required.
Is that the only thing filesystem plugin's used for in plugin-mdx? If yes - I could send a PR with docs 🤔

**Comment by jonniebigodes** [comment]
@yamalight and you're right. After the `.mdx` files are loaded they are transformed and the appropriate GraphQl nodes are created. But before that the actual content needs to be loaded and that's handled by `gatsby-source-filesystem`. Doing a quick check on the documentation available looks like there's no mention to that requirement in order for the plugin to work and that minor oversight might lead to issues like the one you're experiencing. If you're willing read the [documentation on how to contribute](https://www.gatsbyjs.org/contributing/how-to-contribute/), make the necessary adjustments and submit a PR. Sounds good?

**Comment by yamalight** [comment]
@jonniebigodes sounds good! expect a PR soon-ish :)

**Comment by jonniebigodes** [comment]
@yamalight forgot one thing. Mention this issue when you submit the PR if you don't mind.

**Comment by yamalight** [comment]
@jonniebigodes got it, will do!

**Comment by yamalight** [comment]
@jonniebigodes And PR is up under #15604 
Let me know if you want any changes.

**Comment by muescha** [comment]
@jonniebigodes make it sense that the mdx plugin itself also check if the `gatsby-source-filesystem` is activated?

**Comment by johno** [comment]
> @jonniebigodes make it sense that the mdx plugin itself also check if the gatsby-source-filesystem is activated?

Nodes that `gatsby-plugin-mdx` process can be sourced from places other than the filesystem (like a CMS), so we can't necessarily assume that `gatsby-source-filesystem` needs to be present.

**Comment by lannonbr** [comment]
Resolved in #15604
